dreem/models/embedding.py

```diff
@@ -3,15 +3,127 @@
+            the rotation angles
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        # max_seq_len: int,
+        base: int = 10000,
+    ) -> None:
+        super().__init__()
+        self.dim = dim
+        self.base = base
+        # self.max_seq_len = max_seq_len
+        self._rope_init()
+
+    # We need to explicitly define reset_parameters for FSDP initialization, see
+    # https://github.com/pytorch/pytorch/blob/797d4fbdf423dd9320ebe383fb57ffb1135c4a99/torch/distributed/fsdp/_init_utils.py#L885
+    def reset_parameters(self):
+        self._rope_init()
+
+    def _rope_init(self):
+        theta = 1.0 / (
+            self.base
+            ** (torch.arange(0, self.dim, 2)[: (self.dim // 2)].float() / self.dim)
+        )
+        self.register_buffer("theta", theta, persistent=False)
+
+    def build_rope_cache(self, max_seq_len: int) -> None:
+        # Create position indexes `[0, 1, ..., max_seq_len - 1]`
+
+        seq_idx = torch.arange(
+            max_seq_len, dtype=self.theta.dtype, device=self.theta.device
+        )
+
+        # Outer product of theta and position index; output tensor has
+        # a shape of [max_seq_len, dim // 2]
+        idx_theta = torch.einsum("i, j -> ij", seq_idx, self.theta).float()
+
+        # cache includes both the cos and sin components and so the output shape is
+        # [max_seq_len, dim // 2, 2]
+        cache = torch.stack([torch.cos(idx_theta), torch.sin(idx_theta)], dim=-1)
+        self.register_buffer("cache", cache, persistent=False)
+
+    def forward(self, x: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [b, s, n_h, h_d]
+            input_pos (Optional[Tensor]): Optional tensor which contains the position ids
+                of each token. During training, this is used to indicate the positions
+                of each token relative to its sample when packed, shape [b, s].
+                During inference, this indicates the position of the current token.
+                If none, assume the index of the token is its position id. Default is None.
+
+        Returns:
+            Tensor: output tensor with RoPE applied
+
+        Notation used for tensor shapes:
+            - b: batch size
+            - s: sequence length
+            - n_h: num heads
+            - h_d: head dim
 
+        """
+        # input tensor has shape [b, s, n_h, h_d]
+        seq_len = x.size(1)
+
+        # create the lookup array based on how many instances there are
+        # max(101, seq_len) is for positional vs temporal; pos can only have idx up to
+        # 100 since it's a fraction of [0,1]*100. temp is from [0, clip_len]; since clip_len
+        # not available, we use # of instances from input x; this is always >= clip_len
+        self.build_rope_cache(max(101, seq_len)) # registers cache
```

Comment:
Maybe I'm misunderstanding, but then shouldn't this be `min`?

Also, is the assumption that the number of instances >= `clip_len` strictly true? What if we have frames with no detections?